pkg/controller/machine/machine.go

```diff
@@ -888,8 +911,31 @@ func (r *Reconciler) getNode(instance instance.Instance, provider providerconfig
 				return node.DeepCopy(), true, nil
 			}
 		}
+		// If we were unable to find Node by ProviderID, fallback to IP address matching.
+		// This usually happens if there's no CCM deployed in the cluster.
+		//
+		// This mechanism is not always reliable, as providers reuse the IP addresses after
+		// some time.
+		//
+		// If we rollout a Machine, it can happen that a new instance has the same
+		// IP addresses as the instance that has just been deleted. If machine-controller
+		// processes the new Machine before removing the old Machine and the corresponding
+		// Node object, machine-controller could update the NodeOwner label on the old Node
+		// object to point to the Machine that just got created, as IP addresses would match.
+		// This causes machine-controller to fail to delete the old Node object, which could
+		// then cause cluster stability issues in some cases.
 		for _, nodeAddress := range node.Status.Addresses {
 			for instanceAddress := range instance.Addresses() {
+				// Hetzner is known to assign old IP addresses to new instances right after
+				// the old instance has been deleted. On Hetzner, the described issue happens much
+				// more often than for other providers, and as we know that the node name will
+				// always match the instance name, we can check that those two names match.
+				//
+				// TODO: Ideally, we should check this for other providers too, but it's not the
+				// case that node name and instance name will be same for all providers (e.g. AWS).
```

Comment:
As far as I've seen, AWS is actually the only provider that has a different node and instance names. However, I'm not sure how to reliably confirm this.

Reply:
I think it's ok until we have users reporting issues. Here we are covering a corner case that should only happen in very rare circumstances. Normally CCM is deployed and `providerID` is assigned.